bloom-housing/bloom
# feature: preferences on multiple pages

# Pull Request Template

## Issue

Addresses #1236 

## Description

This change allows preferences to be split into a one-per-page flow. I added a new column in the preference table `page` that starts at 1, which indicates which page a preference should live on. The data migration auto-adds every preference to have a page of 1 if it wasn't listed.

If there are multiple preference pages, the form handles it by staying on one URL/form step and just keeping track of the page using state. Adjustments had to be made in order to handle the "back" flow properly, both between preference pages and from the following page. My solution to this was to add a `navigatedThroughBack` flag on the `ApplicationConductor` class and to set the flag whenever we hit the back button. It also meant I had to update the `FormBackLink` component so that it doesn't always change the URL (in the case that we have multiple preference pages).

## Type of change


## How Can This Be Tested/Reviewed?

I recommend puling the branch, running db migrations, and trying the normal flow on the 2-preference listing (they will all be one page) to ensure it works that way still (I initially broke this).

Then I recommend re-seeding which will get the 2-preference listing to have the preferences on multiple pages and trying this out.

Move forward and backward, make updates then move throughout the flow, etc.

## Checklist:
